README.md
## flexi_productimage_slider

 An image slider to represents your products..:)

## Features

 - Slide your products with thumbnail 
 - set your thumbnail in any shapes, color, corner radius, size , alignment etc. as per your requirements
 - perform your action on click on image

<table>
   <tr>
      <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_productimage_slider/master/assets/1.png">
      </td>   
        <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_productimage_slider/master/assets/2.png">
      </td> 
        <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_productimage_slider/master/assets/3.png">
      </td> 
    </tr>
  <tr>
      <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_productimage_slider/master/assets/4.png">
      </td>   
        <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_productimage_slider/master/assets/5.png">
      </td> 
        <td>
         <img width="250px" src="https://raw.githubusercontent.com/Dharini17/flexi_productimage_slider/master/assets/6.png">
      </td> 
    </tr> 
</table>

## Installation

  In your `pubspec.yaml` file within your Flutter Project:

```yaml
dependencies:
  flexi_productimage_slider: 1.1.0
```

```dart

import 'package:flexi_productimage_slider/flexi_productimage_slider.dart';

    flexiProductimageSlider(
    
        //required fields
        arrayImages: const [
        "https://i.ibb.co/ZLFHX3F/1.png",
        "https://i.ibb.co/JKJvs5S/2.png",
        "https://i.ibb.co/LCzV7b3/3.png",
        "https://i.ibb.co/L8JHn1L/4.png",
        "https://i.ibb.co/7RWNCXH/5.png",
        "https://i.ibb.co/bBsh5Pm/6.png",
        ],
        
        // optional fields
        
        //set where you want to set your thumbnail
        sliderStyle: SliderStyle.nextToSlider,//.overSlider, .nextToSlider
        
        // set you slider height like 1.0,1.5,2.0 etc...
        aspectRatio: 0.8,
        
        //set content mode of image
        boxFit: BoxFit.cover,

        //set this if you want to set any default image index when it loads
        selectedImagePosition: 0,
        
        //set your thumbnail alignment 
        thumbnailAlignment: ThumbnailAlignment.bottom,//.right , .left, .bottom

        //set how selected thumbnail border displayed
        thumbnailBorderType: ThumbnailBorderType.all,//.bottom, .all

        //set selected thumbnail border width
        thumbnailBorderWidth: 1.5,//double value
        
        //set thumbnail corner radius
        thumbnailBorderRadius: 10,
        
        //set your thumbnail height & width
        //NOTE : if you set ThumbnailShape.circle then set thumbnail width height same
        thumbnailWidth: 50,
        thumbnailHeight: 65,
        
        //set color of current image thumbnail border
        thumbnailBorderColor: Colors.blue,
        
        //make you action when user click on image
        onTap: (index){
           print("selected index : $index");

           //for zooming effect on click use gallery_zoom_slides
           //https://pub.dev/packages/gallery_zoom_slides
           Navigator.push(context, MaterialPageRoute(builder: (context)=>
                galleryZoomSlides(                
                        //required fields
                        arrayImages: arrayImages,
                        
                        //Optional fields
                        zoomTheme: ZoomTheme.theme3,//.theme1, .theme2, .theme3
                        selectedImagePosition: index,
                        selectedThumbnailColor: Colors.blue,
                  )
                  )
                );
           
        },
    ),

...


```
NOTE : Images taken only for demo purpose


CHANGELOG.md
## 1.0.9 & 1.1.0
- Technical bug Fixing

## 1.0.8
- Set default image position 

## 1.0.7
- Set thumbnail image in any shape without limitation 

## 1.0.6
- Now we can set thumbnail border radius

## 1.0.5
- More thumbnail customization functions & onTap access on image

## 1.0.4 

- Now differentiate selected thumbnail

## 1.0.3

- Bug fixes

## 1.0.2

- Now thumbnail can also display in rectangle

## 1.0.1

- UI changes

## 1.0.0

- It gives you 9 different options with styles to represent your product images


test/flexi_productimage_slider_test.dart
import 'package:flexi_productimage_slider/flexi_productimage_slider.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    final awesome = Awesome();

    setUp(() {
      // Additional setup goes here.
    });

    test('First Test', () {
      expect(awesome.isAwesome, isTrue);
    });
  });
}


lib/src/ZoomView.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:gallery_zoom_slides/gallery_zoom_slides.dart';

class ZoomView extends StatelessWidget {

  int currentIndex;
  List<String> arrayImages;
  ZoomView({required this.arrayImages,required this.currentIndex});

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(

        body:  gallery_zoom_slides(
            arrayImages: arrayImages,
            currentIndex: currentIndex
        )

    );
  }
}

lib/src/flexi_productimage_slider_base.dart
import 'package:flutter/material.dart';
import 'package:cached_network_image/cached_network_image.dart';

enum ThumbnailAlignment { left, right, bottom }
enum ThumbnailShape { circle, custom }
enum ThumbnailBorderType { all, bottom }
enum SliderStyle { overSlider, nextToSlider, }

class flexiProductimageSlider extends StatelessWidget{

  final List<String> arrayImages;
  final double? aspectRatio;
  final BoxFit? boxFit;
  final ThumbnailAlignment? thumbnailAlignment;
  final double? thumbnailWidth;
  final double? thumbnailHeight;
  final Color? thumbnailBorderColor;
  final ThumbnailBorderType? thumbnailBorderType;
  final double? thumbnailBorderRadius;
  final double? thumbnailBorderWidth;
  final SliderStyle? sliderStyle;
  final Function? onTap;
  final int? selectedImagePosition;

  flexiProductimageSlider({
    required this.arrayImages,
    this.aspectRatio = 16/9,
    this.boxFit = BoxFit.fill,
    this.thumbnailAlignment = ThumbnailAlignment.left,
    this.thumbnailWidth = 40,
    this.thumbnailHeight = 40,
    this.thumbnailBorderColor = Colors.blueGrey,
    this.thumbnailBorderType = ThumbnailBorderType.all,
    this.thumbnailBorderRadius = 0,
    this.thumbnailBorderWidth = 1,
    this.sliderStyle,
    this.onTap,
    this.selectedImagePosition = 0,
  });

  double paddingOfBorder = 3;
  PageController pageController = PageController(initialPage: 0,);
  ValueNotifier<int> currentPageViewPage = ValueNotifier(0);

  @override
  Widget build(BuildContext context) {

    currentPageViewPage = ValueNotifier(selectedImagePosition!);

    Future.delayed(const Duration(milliseconds: 3), () {
      pageController.jumpToPage(selectedImagePosition!);
    });

    return  buildContent();
  }

  Widget buildContent(){

    switch (sliderStyle){

      case SliderStyle.overSlider: return buildTheme1();

      case SliderStyle.nextToSlider: return buildTheme2();

      default: return buildTheme1();

    }

  }

  Widget buildTheme1(){
    return AspectRatio( aspectRatio: aspectRatio!,
        child:Stack(
          children: [

            Positioned.fill(
                child: buildImageSlider()
            ),

            thumbnailAlignment == ThumbnailAlignment.bottom ?

            Positioned(
                left: 0,bottom: 5,right: 0,
                child: buildThumbnail(isVertical: false)
            ) :

            thumbnailAlignment == ThumbnailAlignment.left ?

            Positioned(
                top: 5,left: 10,bottom: 5,
                child: buildThumbnail(isVertical: true)
            )
                :
            Positioned(
                top: 5,right: 10,bottom: 5,
                child: buildThumbnail(isVertical: true)
            )


          ],
        ));
  }

  Widget buildTheme2(){
    return

      thumbnailAlignment == ThumbnailAlignment.bottom ?
      Column(
        children: [

          AspectRatio( aspectRatio: aspectRatio!,
            child:buildImageSlider(),
          ),


          Padding(padding: EdgeInsets.only(top: 5,bottom: 5),
            child: buildThumbnail(isVertical: false),
          )

        ],
      ):
      AspectRatio( aspectRatio: aspectRatio!,
          child:
          Row(
            children: [

              thumbnailAlignment == ThumbnailAlignment.left ?
              Padding(padding: EdgeInsets.only(left: 5,right: 5),
                child: buildThumbnail(isVertical: true),
              ) : const SizedBox(),

              Expanded(
                child:  buildImageSlider(),
              ),

              thumbnailAlignment == ThumbnailAlignment.right ?
              Padding(padding: EdgeInsets.only(left: 5,right: 5),
                child: buildThumbnail(isVertical: true),
              ) : const SizedBox(),

            ],
          ));
  }

  Widget buildImageSlider(){
    return PageView.builder(
      controller:pageController,
      onPageChanged: (int currentPage){

        currentPageViewPage.value = currentPage;
        currentPageViewPage.notifyListeners();

      },
      itemCount: arrayImages.length,
      itemBuilder: (context, index) {

        return

          InkWell(
            onTap: ()=> onTap!(index),
            child: funcDisplayImage(arrayImages[index],boxFit!),
          );

      },
    );
  }

  Widget buildThumbnail({required bool isVertical}){
    return Container(
      width: isVertical ? thumbnailWidth! + paddingOfBorder + 10 : double.infinity,
      height: isVertical ? double.infinity : thumbnailHeight! + paddingOfBorder + 10,
      child:  ValueListenableBuilder(
          valueListenable: currentPageViewPage,
          builder: (BuildContext context, currIndexValue, child) {

            return ListView.builder(
                    scrollDirection: isVertical ? Axis.vertical : Axis.horizontal,
                    itemCount: arrayImages.length,
                    itemBuilder: (context, index) {
                      return Padding(padding: isVertical ? EdgeInsets.zero : EdgeInsets.only(left: 5),
                        child: Column(
                          children: [
                            InkWell(
                                onTap: (){
                                 currentPageViewPage.value = index;
                                  currentPageViewPage.notifyListeners();
                                  pageController.jumpToPage(index);

                                },
                                child:

                               Container(
                                        height:thumbnailHeight! + paddingOfBorder,
                                        width: thumbnailWidth! + paddingOfBorder,
                                        decoration: thumbnailBorderType == ThumbnailBorderType.all ?

                                        BoxDecoration(
                                          borderRadius: BorderRadius.circular(thumbnailBorderRadius!),
                                          border: currIndexValue != index ? Border.all(width: thumbnailBorderWidth! ,color: Colors.transparent) :
                                          Border.all(width: thumbnailBorderWidth!,color: thumbnailBorderColor!),
                                        )
                                            :
                                        BoxDecoration(
                                          border:
                                          Border(
                                            bottom: BorderSide(width: thumbnailBorderWidth!,color:currIndexValue != index ? Colors.transparent : thumbnailBorderColor!),
                                          ),
                                        ),
                                        child: Padding(
                                          padding: EdgeInsets.all(3),
                                          child:ClipRRect(
                                            borderRadius: BorderRadius.circular(thumbnailBorderRadius!),
                                            child:  funcDisplayImage(arrayImages[index],boxFit!),
                                          ),
                                        ),
                                      )

                            ),
                            const SizedBox(height: 10,)
                          ],
                        ),
                      );
                    },
              );
          })
    );
  }

  //display network image
  static CachedNetworkImage funcDisplayImage(String strImageURL, BoxFit varBoxType) {

    return CachedNetworkImage(
      imageUrl: strImageURL,
      imageBuilder: (context, imageProvider) => Container(
        decoration: BoxDecoration(
          image: DecorationImage(
            image: imageProvider,
            fit: varBoxType,
          ),
        ),
      ),
      errorWidget: (context, url, error) => Center(
        child: Icon(Icons.image_outlined,color: Colors.grey.shade400,),
      ),
    );
  }
}

//----------------------

// import 'package:cached_network_image/cached_network_image.dart';
// import 'package:flutter/material.dart';
//
// enum ThumbnailAlignment { left, right, bottom }
//
// enum ThumbnailShape { circle, custom }
//
// enum SliderStyle { overSlider, nextToSlider, }
//
// class flexiProductimageSlider extends StatelessWidget{
//
//   final List<String> arrayImages;
//   final double? aspectRatio;
//   final BoxFit? boxFit;
//   final ThumbnailAlignment? thumbnailAlignment;
//   final ThumbnailShape? thumbnailShape;
//   final double? thumbnailWidth;
//   final double? thumbnailHeight;
//   final Color? thumbnailBorderColor;
//   final double? thumbnailBorderRadius;
//   final SliderStyle? sliderStyle;
//   final Function? onTap;
//
//   flexiProductimageSlider({
//     required this.arrayImages,
//     this.aspectRatio = 16/9,
//     this.boxFit = BoxFit.fill,
//     this.thumbnailAlignment = ThumbnailAlignment.left,
//     this.thumbnailShape = ThumbnailShape.custom,
//     this.thumbnailWidth = 40,
//     this.thumbnailHeight = 40,
//     this.thumbnailBorderColor = Colors.blueGrey,
//     this.thumbnailBorderRadius = 0.0,
//     this.sliderStyle,
//     this.onTap,
//   });
//
//   PageController pageController = PageController(initialPage: 0,);
//   ValueNotifier<int> currentPageViewPage = ValueNotifier(0);
//
//   @override
//   Widget build(BuildContext context) {
//
//     return  buildContent();
//   }
//
//   Widget buildContent(){
//
//     switch (sliderStyle){
//
//       case SliderStyle.overSlider: return buildTheme1();
//
//       case SliderStyle.nextToSlider: return buildTheme2();
//
//       default: return buildTheme1();
//
//     }
//
//   }
//
//   Widget buildTheme1(){
//     return AspectRatio( aspectRatio: aspectRatio!,
//               child:Stack(
//       children: [
//
//         Positioned.fill(
//             child: buildImageSlider()
//         ),
//
//         thumbnailAlignment == ThumbnailAlignment.bottom ?
//
//         Positioned(
//             left: 0,bottom: 5,right: 0,
//             child: buildThumbnail(isVertical: false)
//         ) :
//
//         thumbnailAlignment == ThumbnailAlignment.left ?
//
//         Positioned(
//             top: 5,left: 10,bottom: 5,
//             child: buildThumbnail(isVertical: true)
//         )
//             :
//         Positioned(
//             top: 5,right: 10,bottom: 5,
//             child: buildThumbnail(isVertical: true)
//         )
//
//
//       ],
//     ));
//   }
//
//   Widget buildTheme2(){
//     return
//
//       thumbnailAlignment == ThumbnailAlignment.bottom ?
//     Column(
//       children: [
//
//       AspectRatio( aspectRatio: aspectRatio!,
//       child:buildImageSlider(),
//       ),
//
//
//         Padding(padding: EdgeInsets.only(top: 5,bottom: 5),
//           child: buildThumbnail(isVertical: false),
//         )
//
//       ],
//     ):
//       AspectRatio( aspectRatio: aspectRatio!,
//           child:
//       Row(
//         children: [
//
//           thumbnailAlignment == ThumbnailAlignment.left ?
//           Padding(padding: EdgeInsets.only(left: 5,right: 5),
//             child: buildThumbnail(isVertical: true),
//           ) : const SizedBox(),
//
//           Expanded(
//               child:  buildImageSlider(),
//           ),
//
//           thumbnailAlignment == ThumbnailAlignment.right ?
//           Padding(padding: EdgeInsets.only(left: 5,right: 5),
//             child: buildThumbnail(isVertical: true),
//           ) : const SizedBox(),
//
//         ],
//       ));
//   }
//
//   Widget buildImageSlider(){
//     return PageView.builder(
//       controller:pageController,
//       onPageChanged: (int currentPage){
//
//         currentPageViewPage.value = currentPage;
//         currentPageViewPage.notifyListeners();
//
//       },
//       itemCount: arrayImages.length,
//       itemBuilder: (context, index) {
//
//         return
//
//           InkWell(
//             onTap: ()=> onTap!(index),
//             child: funcDisplayImage(arrayImages[index],boxFit!),
//           );
//
//       },
//     );
//   }
//
//   Widget buildThumbnail({required bool isVertical}){
//     return Container(
//        width: isVertical ? thumbnailWidth! + 10 : double.infinity,
//       height: isVertical ? double.infinity : thumbnailHeight! + 10,
//       child: ListView.builder(
//
//         scrollDirection: isVertical ? Axis.vertical : Axis.horizontal,
//         itemCount: arrayImages.length,
//         itemBuilder: (context, index) {
//           return Padding(padding: isVertical ? EdgeInsets.zero : EdgeInsets.only(left: 10),
//             child: Column(
//               children: [
//                 InkWell(
//                   onTap: (){
//
//                     pageController.animateToPage(index, duration: Duration(milliseconds: 400), curve: Curves.linear);
//                     currentPageViewPage.value = index;
//                     currentPageViewPage.notifyListeners();
//
//                   },
//                   child:
//
//                   ValueListenableBuilder(
//                       valueListenable: currentPageViewPage,
//                       builder: (BuildContext context, currIndexValue, child) {
//
//                         return Container(
//                           height:thumbnailHeight,
//                           width: thumbnailWidth,
//                           decoration: BoxDecoration(
//                               borderRadius: BorderRadius.circular( thumbnailShape == ThumbnailShape.circle ?  (thumbnailHeight!/2) : thumbnailBorderRadius!),
//                               border: currIndexValue != index ? Border.all(width: 1 ,color: Colors.transparent) : Border.all(width: 1,color: thumbnailBorderColor!),
//                           ),
//                           child: Padding(
//                             padding: EdgeInsets.all(thumbnailShape == ThumbnailShape.circle ? 0 : 5),
//                             child:ClipRRect(
//                             borderRadius: BorderRadius.circular( thumbnailShape == ThumbnailShape.circle ? (thumbnailHeight!/2) : thumbnailBorderRadius!),
//                             child:  funcDisplayImage(arrayImages[index],boxFit!),
//                             ),
//                           ),
//                         );
//                       })
//
//                 ),
//                 const SizedBox(height: 10,)
//               ],
//             ),
//           );
//         },
//
//       ),
//     );
//   }
//
//   //display network image
//   static CachedNetworkImage funcDisplayImage(String strImageURL, BoxFit varBoxType) {
//
//     return CachedNetworkImage(
//       imageUrl: strImageURL,
//       imageBuilder: (context, imageProvider) => Container(
//         decoration: BoxDecoration(
//           image: DecorationImage(
//             image: imageProvider,
//             fit: varBoxType,
//           ),
//         ),
//       ),
//       errorWidget: (context, url, error) => Center(
//         child: Icon(Icons.image_outlined,color: Colors.grey.shade400,),
//       ),
//     );
//   }
// }


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flexi_productimage_slider/flexi_productimage_slider.dart';

void main() async {

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {

    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'flexi_productimage_slider',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'flexi_productimage_slider'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {


  @override
  void initState() {
    super.initState();

  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body:  SingleChildScrollView(
        child: Column(

          children: <Widget>[

            const SizedBox(height: 10,),

            flexiProductimageSlider(

              //required fields
                arrayImages: const [
                "https://i.ibb.co/ZLFHX3F/1.png",
                "https://i.ibb.co/JKJvs5S/2.png",
                "https://i.ibb.co/LCzV7b3/3.png",
                "https://i.ibb.co/L8JHn1L/4.png",
                "https://i.ibb.co/7RWNCXH/5.png",
                "https://i.ibb.co/bBsh5Pm/6.png",
                ],

                // optional fields

                //set where you want to set your thumbnail
                sliderStyle: SliderStyle.overSlider,//SliderStyle.nextToSlider

                // set you slider height like 1.0,1.5,2.0 etc...
                aspectRatio: 0.8,

                //set content mode of image
                boxFit: BoxFit.cover,

                //set your thumbnail alignment & shape
                thumbnailAlignment: ThumbnailAlignment.right,//ThumbnailAlignment.right , ThumbnailAlignment.bottom
                thumbnailShape: ThumbnailShape.circle,//ThumbnailShape.circle

               //set your thumbnail height & width
                //NOTE : if you set ThumbnailShape.circle then set thumbnail width height same
                thumbnailWidth: 50,
                thumbnailHeight: 65,

                //set color of current image thumbnail border
                thumbnailBorderColor: Colors.deepOrangeAccent,

                //make you action when user click on image
                onTap: (index){
                  print("selected index : $index");
                },
             ),


          ],
        ),
      ),
      // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
